### PR TITLE
Add custom action to handmade products grid

### PR DESCRIPTION
**Pull request by jamesricky**
We need to manually set the width of the actions column, depending on the number of `IconButton` components shown, 32px per `IconButton`.  DataGrid columns do not scale automatically, depending on their content. 

We cannot use `getActions` from the [special property types](https://mui.com/x/react-data-grid/column-definition/#special-properties) for this:

- Every item returned by `getActions` must be the `GridActionsCellItem` component directly. This would not allow us to wrap the item with a `Tooltip` or add a Fragment that contains the item so we can add a `Dialog` that opens when clicking the item. 

- Additionally, `CrudContextMenu` is not currently implemented with `GridActionsCellItem` so it cannot be used inside `getActions` in it's current state. Simply replacing the items with `GridActionsCellItem` would not work, as `getActions` requires an array of items and cannot handle multiple items inside one element or fragment. 

---

## PR Checklist

-   [x] Link to the respective task if one exists: SVK-211
-   [x] Provide screenshots/screencasts if the change contains visual changes

https://github.com/vivid-planet/comet/assets/6264317/e500c050-0474-404f-b4f3-5f22e9d6df23

### CODE REVIEW

**Comment by nsams** [comment]
> Every item returned by getActions must be the GridActionsCellItem component directly. This would not allow us to wrap the item with a Tooltip or add a Fragment that contains the item so we can add a Dialog that opens when clicking the item.

this is not correct, at least in mui-x v7 there is an example in the [docs](https://mui.com/x/react-data-grid/column-definition/#special-properties) that does exaclty that


> Additionally, CrudContextMenu is not currently implemented with GridActionsCellItem so it cannot be used inside getActions in it's current state. Simply replacing the items with GridActionsCellItem would not work, as getActions requires an array of items and cannot handle multiple items inside one element or fragment.

we could change that, this should not be an argument.


(I don't want to say that we should use `type="actions"`)

**Comment by jamesricky** [comment]
> > Every item returned by getActions must be the GridActionsCellItem component directly. This would not allow us to wrap the item with a Tooltip or add a Fragment that contains the item so we can add a Dialog that opens when clicking the item.
> 
> this is not correct, at least in mui-x v7 there is an example in the [docs](https://mui.com/x/react-data-grid/column-definition/#special-properties) that does exaclty that

Where exactly? I don't see it 😅 

> > Additionally, CrudContextMenu is not currently implemented with GridActionsCellItem so it cannot be used inside getActions in it's current state. Simply replacing the items with GridActionsCellItem would not work, as getActions requires an array of items and cannot handle multiple items inside one element or fragment.
> 
> we could change that, this should not be an argument.

True.

**Comment by nsams** [comment]
Bildschirmfoto vom 2024-05-13 09-11-53

**Comment by nsams** [comment]
> came to the conclusion that there is no real benefit to using it, over our current approach

imho the main benefit would be that we are using a standard DataGrid feature instead of creating our own. If both work well, I'd always choose the standard way.

why?

- less code to maintain on our side
- devs can consult mui docs/examples (the more we move away from mui, the less mui docs are even opened by devs)

> It would however add the necessity for a few workarounds regarding styling and typing.

If those are some theming stuff internally, I would accept that. If it results in an not-so-nice API for devs, then it's an argument against.

**Comment by jamesricky** [comment]
@nsams all good points, I'll create a separate PR to try this out so we can see the pros/cons more easily.

**Comment by jamesricky** [comment]
@nsams I tried implementing this with `getActions()`: https://github.com/vivid-planet/comet/commit/a8d15f8aab5dc15038e1edf009258d54454e73a4

Other than styling and possible types we would have to override, the `GridActionsCellItem` component seems to be quite buggy. I was not able to get the delete-item of the CRUD actions to work. Because of the fragment around the item, the item for some reason replaces all other items and is rendered outside of the ellipsis menu. 

There's also a bunch of TS problems, such as `getActions` not allowing items to be returned with custom props, the `onResize` props being required, even though it should be `undefined`. 

Also items returned in `getActions` would need a key. 

In conclusion, I don't think it's worth the effort of refactoring/restyling everything and adding a bunch of typescript workarounds, just to have a slightly different API. Especially because of the buggy behavior.

**Review by nsams — APPROVED** [review]

File: demo/admin/src/products/ProductsGrid.tsx [hunk 1]
```diff
@@ -139,9 +140,11 @@ export function ProductsGrid() {
             headerName: "",
             sortable: false,
             filterable: false,
+            width: 106,
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply demo/admin/src/products/ProductsGrid.tsx:143]: See the PR description 🙂 

> DataGrid columns do not scale automatically, depending on their content.

[the remark this answers is not in the document]
Comment (johnnyomair) [reply demo/admin/src/products/ProductsGrid.tsx:143]: The Admin Generator should make this multiplication by default (X * 32)

**Review by johnnyomair — APPROVED** [review]
